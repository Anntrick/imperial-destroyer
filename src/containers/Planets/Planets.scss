@import "../../themes/colors.scss";

.Planets {
    color: $text-color;
    padding: 2em;

    ul {
        display: flex;
        flex-wrap: wrap;
        
    }
}