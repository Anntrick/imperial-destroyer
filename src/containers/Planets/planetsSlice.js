import { createSlice } from "@reduxjs/toolkit"

const initialState = {
    items: [],
    error: ''
}

const planetsSlice = createSlice({
    name: "planets",
    initialState,
    reducers: {
        addPlanets(state, action) {
            return {
                ...state,
                items: action.payload
            }
        },
        markAsConquered(state, action) {
            const planetIndex = state.items.findIndex(planet => planet.name === action.payload)
            if (planetIndex !== -1) {
                return {
                    ...state,
                    items: [
                        ...state.items,
                        state.items[planetIndex].conquered = true
                    ]
                }
            }

        },
        markAsDeleted(state, action) {
            const planetIndex = state.items.findIndex(planet => planet.name === action.payload)
            if (planetIndex !== -1) {
                return {
                    ...state,
                    items: [
                        ...state.items,
                        state.items[planetIndex].deleted = true
                    ]
                }
            }
        },
        sortPlanets(state, action) {
            const { field, direction } = action.payload;
            state.items.sort((a, b) => {
                if (a[field] < b[field])
                    return -1
                if (a[field] > b[field])
                    return 1
                return 
            })
        },
        setError(state, action) {
            return {
                ...state,
                error: action.payload
            }
        },
    },
})

export const { addPlanets, markAsConquered, markAsDeleted, sortPlanets, setError } = planetsSlice.actions

export const planetsData = (state) => state.planetsSlice
export default planetsSlice.reducer