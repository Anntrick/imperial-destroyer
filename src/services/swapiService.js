import axios from "axios"

export const SwapiService = {}
const apiUrl = 'https://swapi.tech/api/'

SwapiService.getPlanets = async () => {
  try {
    const response = await axios.get(`${apiUrl}planets/`)
    
    return await SwapiService.addData(response.data.results)
  } catch (error) {
    console.error(error)
    return error
  }
}

SwapiService.getStarships = async () => {
  try {
    const response = await axios.get(`${apiUrl}starships/`)
    
    return await SwapiService.addData(response.data.results)
  } catch (error) {
    console.error(error)
    return error
  }
}

SwapiService.addData = async (results) => {
  console.log(results)
  let data = []
  for (let item of results) {
      const itemData = await axios.get(item.url)
      
      data.push(itemData.data.result.properties)
  }
  return data
}