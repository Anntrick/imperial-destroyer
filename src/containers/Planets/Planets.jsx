import React, { useEffect, useState } from "react"
import { useSelector, useDispatch } from "react-redux"
import { addPlanets, markAsConquered, markAsDeleted, sortPlanets, setError, planetsData } from "./planetsSlice"
import { SwapiService } from '../../services/swapiService'

import './Planets.scss'
import { Loading } from "../../components/Loading/Loading"
import { ItemCard } from "../../components/ItemCard/ItemCard"

export const Planets = () => {
  const dispatch = useDispatch()
  const planets = useSelector(planetsData)

  const [searchTerm, setSearchTerm] = useState("")
  const [sortBy, setSortBy] = useState("")

  useEffect(() => {
    getPlanets()

  }, [])

  const getPlanets = async () => {
    try {
      const data = await SwapiService.getPlanets()
      dispatch(addPlanets(data))
    } catch (error) {
      dispatch(setError(error))
    }

  }

  const handleSort = (field) => {
    setSortBy(field)
    dispatch(sortPlanets({ field }))
  }

  const filteredPlanets = planets?.items?.filter((planet) =>
    planet.name.toLowerCase().includes(searchTerm.toLowerCase())
  )

  return (
    <div className="Planets">
      <h2>Planets</h2>
      {planets?.length === 0 &&
        <Loading />
      }


      <input type="text" value={searchTerm} onChange={(e) => setSearchTerm(e.target.value)} placeholder="Search by planet name" />
      <button onClick={() => handleSort("name")}>Sort by Name</button>
      <ul>
        {filteredPlanets?.map((planet) => (
          <ItemCard data={planet} key={planet.name} />
        ))}


      </ul>
    </div>


  )

}