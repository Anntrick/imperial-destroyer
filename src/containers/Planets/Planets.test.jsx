import React from 'react'
import { render, fireEvent, getByText, act } from '@testing-library/react'
import { Planets } from './Planets'
import { Provider, useDispatch, useSelector } from 'react-redux'
import { planetsSlice, addPlanets } from './planetsSlice'
import { combineReducers, configureStore } from '@reduxjs/toolkit'
import '@testing-library/jest-dom'
import { SwapiService } from '../../services/swapiService'
import { vi } from 'vitest'

vi.mock('react-redux', async () => {
    const actual = await vi.importActual('react-redux')

    return {
        ...actual,
        useSelector: vi.fn().mockReturnValue([{ name: 'Tatooine', population: 200000 }, { name: 'Alderaan', population: 200000000 }]),
        useDispatch: vi.fn()
    }
})

vi.mock(SwapiService)

const planets = [
    { name: 'Alderaan', population: '2000000000', climate: 'temperate', terrain: 'grasslands, mountains' },
    { name: 'Tatooine', population: '200000', climate: 'arid', terrain: 'desert' },
    { name: 'Hoth', population: 'unknown', climate: 'frozen', terrain: 'tundra, ice caves' },

]

describe('Planets component', () => {
    let store, getPlanetsSpy

    beforeEach(() => {
        getPlanetsSpy = vi.spyOn(SwapiService, "getPlanets").mockImplementation(() => Promise.resolve(planets))
        store = configureStore({
            reducer: combineReducers({
                planetsSlice
            })
        })
    })

    afterEach(() => {
        vi.clearAllMocks()
    })
    
    it('Should render the component correctly', () => {
        const { getByText } = render(
            <Provider store={store}>
                <Planets />
            </Provider>
        )
        expect(getByText("Planets")).toBeInTheDocument()
    })

    it('Should render Loading component when loading is true', () => {
        const { getByTestId } = render(
            <Provider store={store}>
                <Planets />
            </Provider>
        )
        expect(getByTestId('loading')).toBeInTheDocument()
    })

    it('Should render input and button', () => {
        const { getByPlaceholderText, getByText } = render(
            <Provider store={store}>
                <Planets />
            </Provider>
        )
        expect(getByPlaceholderText('Search by planet name')).toBeInTheDocument()
        expect(getByText('Sort by Name')).toBeInTheDocument()
    })

    it('should call dispatch with the API response', async () => {
        store.dispatch = vi.fn()

        await act(async () => {
            render(<Planets />)
        })

        /*expect(dispatch).toHaveBeenCalledWith(addPlanets([
            { name: 'Alderaan', population: '2000000000', climate: 'temperate', terrain: 'grasslands, mountains' },
            { name: 'Tatooine', population: '200000', climate: 'arid', terrain: 'desert' },
            { name: 'Hoth', population: 'unknown', climate: 'frozen', terrain: 'tundra, ice caves' },
        ]))*/
    })

    it('should render ItemCard with correct data', () => {
        const dispatch = vi.fn()
        const { getByText, getByTestId } = render(
            <Provider store={store}>
                <Planets />
            </Provider>
        )
        const mockUseSelector = vi.fn().mockReturnValue(planets)
      //  vi.spyOn(ReactRedux, "useSelector").mockImplementation(mockUseSelector)
       // vi.spyOn(ReactRedux, "useDispatch").mockReturnValue(dispatch)
        expect(useSelector).toHaveBeenCalled()
        //expect(getByText("Tatooine")).toBeInTheDocument()

    })
})




/* 

3. Verificar que se haya manejado correctamente el error de la llamada a la API:

Se debe comprobar que la función "dispatch(setError(error))" se haya llamado en caso de que ocurra un error en la llamada a la API.
4. Verificar que el componente ItemCard se haya renderizado con la información correcta:

Se debe comprobar que se haya pasado un objeto "planet" como "data" a cada componente "ItemCard".
Se debe comprobar que cada componente "ItemCard" se haya renderizado con la información correspondiente de cada planeta.
5. Verificar la búsqueda de planetas:

Se debe comprobar que la búsqueda funcione correctamente.
Se debe comprobar que la búsqueda sea insensible a mayúsculas y minúsculas.
6. Verificar el ordenamiento de planetas:

Se debe comprobar que la función "handleSort" se haya ejecutado al hacer clic en el botón "Sort by Name".
Se debe comprobar que la función "dispatch(sortPlanets({ field }))" se haya llamado con el valor correcto para el campo "field".
7. Verificar la paginación de planetas:

Se debe comprobar que se haya implementado correctamente la paginación de planetas.
Se debe comprobar que el componente "Pagination" se haya renderizado correctamente con los valores correctos para "currentPage", "setCurrentPage", "totalItems" y "itemsPerPage".
 */