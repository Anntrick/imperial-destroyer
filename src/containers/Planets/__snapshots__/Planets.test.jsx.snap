// Vitest Snapshot v1

exports[`Planets component > Should render the component correctly 1`] = `
<DocumentFragment>
  <div
    class="Planets"
  >
    <h2>
      Planets
    </h2>
    <input
      defaultvalue=""
      placeholder="Search by planet name"
      type="text"
    />
    <button>
      Sort by Name
    </button>
    <div
      class="Loading"
    >
      <div
        class="dot data-testid"
      />
    </div>
    <ul />
  </div>
</DocumentFragment>
`;
